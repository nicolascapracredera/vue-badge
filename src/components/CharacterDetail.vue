<template>
    <div v-if="character!==null">
        <h3>Detailed Character Info</h3>
        <div class="card" style="width: 18rem;">
            <img v-if="character.thumbnail" alt="Character Photo" class="card-img-top"
                v-bind:src="character.thumbnail.path + '/portrait_incredible.' + character.thumbnail.extension" />
            <div class="card-body">
                <p>Name: {{ character.name }}</p>
                <p>Appears in {{ character.comics.available }} comics, {{ character.series.available }} series, and {{ character.stories.available }} stories.</p>
                <p>{{ character.name }} was last modified on {{ this.getFormattedDate() }}</p>
            </div>
        </div>
    </div>
</template>

<script>
export default {
    name: 'CharacterDetail',
    props: [ 'character' ],
    data: function() {
        return {
            
        }
    },
    methods: {
        getFormattedDate: function getFormattedDate() {
            var date = new Date(this.character.modified);
            return date.toLocaleDateString();
        }
    }
}
</script>

<style scoped>
#character-detail {
    float: left;
    border: 2px solid black;
    border-radius: 5%;
    padding: 5px;
    background: lightgray;
    width: 30%;
}
</style>