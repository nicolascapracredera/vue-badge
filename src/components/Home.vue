<template>
  <div class="hello">
      <MarvelCharacters></MarvelCharacters>
  </div>
</template>

<script>
import MarvelCharacters from './MarvelCharacters.vue'

export default {
  name: 'Home',
  components: {
    MarvelCharacters
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
h3 {
  margin: 40px 0 0;
}
ul {
  list-style-type: none;
  padding: 0;
}
li {
  display: inline-block;
  margin: 0 10px;
}
a {
  color: #42b983;
}
</style>
