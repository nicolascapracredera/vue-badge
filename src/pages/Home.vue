<template>
  <div>
      <p>Hello, and welcome to the Marvel Character Lookup system. Here, you can search for different characters
        that exist in the Marvel Comics universe and learn more about those characters.</p>
      <p>First, let's create a (mock) account for you, to ensure that you're a real user and so (one day) you can
        save your searches and keep lists for your personal use</p>
    <div>
      <form
  id="form"
>
  <div class="form-group col-4" style="float:left;">
    <label for="name">Name</label>
    <input
      id="name"
      v-model="name"
      type="text"
      name="name"
      class="form-control"
      required
    >
    <div class="valid-feedback">
        Looks good!
      </div>
    <div class="invalid-feedback">
      Your name is required
    </div>
  </div>

  <div class="form-group col-4" style="float:left;">
    <label for="email">Email</label>
    <input
      id="email"
      v-model="email"
      type="text"
      name="email"
      min="0"
      class="form-control"
      required
    >
    <div class="valid-feedback">
      Looks good!
    </div>
    <div class="invalid-feedback">
      {{ emailError }}
    </div>
  </div>

  <div class="form-group col-4" style="float:left;">
    <label for="age">Age {{age}}</label>
    <input
      id="age"
      v-model="age"
      type="range"
      min="1"
      max="99"
      name="age"
      class="custom-range"
      required
    >
    <div class="valid-feedback">
      Looks good!
    </div>
    <div class="invalid-feedback">
      {{ ageError }}
    </div>
  </div>

  <div class="form-group" id="button-container">
    <button
      style="margin-left:50%;"
      type="submit"
      class="btn btn-primary">
      Submit
    </button>
  </div>
</form>
    </div>
  </div>
</template>

<script>

export default {
  name: 'Home',
  data: function() { 
    return {
      name: null,
      email: null,
      age: 1,
      emailError: "",
      ageError: ""
    }
  },
  methods:{
    checkEmail: function () {
      const emailRequiredError = 'Email required.';
      const emailFormatError = 'Email is not in the correct format.';
      const emailRegex = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;

      document.getElementById("email").classList.remove("is-invalid");
      document.getElementById("email").classList.remove("is-valid");

      if (!this.email) {
        this.emailError = emailRequiredError;
        document.getElementById("email").classList.add("is-invalid");
      } else if (!emailRegex.test(this.email)) {
        this.emailError = emailFormatError;
        document.getElementById("email").classList.add("is-invalid");
      } else {
        document.getElementById("email").classList.add("is-valid");
      }
    },
    checkAge: function() {
      const minAgeError = 'You need to be 18 or older to use this service.';

      document.getElementById("age").classList.remove("is-invalid");
      document.getElementById("age").classList.remove("is-valid");

      if (this.age < 18) {
        this.ageError = minAgeError;
        document.getElementById("age").classList.add("is-invalid");
      } else {
        document.getElementById("age").classList.add("is-valid");
      }
    }
  },
  watch: {
    email: function() {
      this.checkEmail();
    },
    age: function() {
      this.checkAge();
    }, 
    name: function() {
      document.getElementById("name").parentElement.classList.add("was-validated");
    }
  }
}
</script>

<style scoped>
#button-container {
  float: left;
  width: 100%;
}
</style>
