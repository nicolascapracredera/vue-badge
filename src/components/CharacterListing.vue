<template>
    <div>
        <p>Total Results: {{ totalAvailable }}</p>
        <ul id="character-listing">
            <li v-for="character in characters" :key="character.id" v-on:click="selectedCharacter = character">
                {{ character.name }}
            </li>
        </ul>
        <CharacterDetail v-bind:character="selectedCharacter"></CharacterDetail>
        <div id="pageButtons" v-if="totalAvailable > 20">
            <button v-on:click="changePage(-1)">Previous Page</button>
            <button v-on:click="changePage(1)">Next Page</button>
        </div>
    </div>
</template>

<script>
import {searchMarvelCharacters} from '../services/marvel'
import {getMarvelCharacters} from '../services/marvel'
import CharacterDetail from './CharacterDetail.vue'

export default {
    name: 'MarvelCharacters',
    props: [ 'searchTerm' ],
    components: {
        CharacterDetail
    },
    data: function() {
        return {
            page: 1,
            characters: [], 
            selectedCharacter: null,
            totalAvailable: 0
        }
    },
    watch: {
        searchTerm: function(val) {
            console.log('char listing ', val);
            this.getCharacters(1);
        }
    },
    methods: {
        getCharacters: function(page) {
            console.log("current search term ", this.searchTerm)
            if (this.searchTerm !== "") {
                searchMarvelCharacters(this.updateCharacters, page, this.searchTerm);
            } else {
                getMarvelCharacters(this.updateCharacters, page);
            }
        },
        updateCharacters: function(incomingCharacters, numResults) {
            this.characters = incomingCharacters;
            this.totalAvailable = numResults;
        },
        changePage: function(amount) {
            if((this.page + amount) >= 1) {
                this.page += amount;
            }
            this.getCharacters(this.page);
        }
    },mounted() {
        this.getCharacters(this.page);
    }
}
</script>