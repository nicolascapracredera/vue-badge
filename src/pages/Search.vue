<template>
    <div id="container">
        <p>Start typing in the box below to search for a Marvel character. Click on a character's name to view details about them.</p>
        <input class="form-control col-4" type="text" v-model="characterSearch" placeholder="Type Character Name">
        <CharacterListing :searchTerm="debouncedSearchTerm"></CharacterListing>
    </div>
</template>

<script>
const _ = require('lodash');
import CharacterListing from '../components/CharacterListing.vue'
require('bootstrap')

export default {
    name: 'Search',
    components: {
        CharacterListing
    },
    data: function() {
        return {
            characterSearch: "",
            character: null,
            debouncedSearchTerm: ""
        }
    },
    watch: {
        characterSearch: _.debounce(function(val) {
            this.debouncedSearchTerm = val;
        }, 1000)
    },
    methods: {
        updateCharacter: function updateCharacter(characters) {
            this.character = characters[0];
        }
    }
}
</script>

<style scoped>
#container {
    margin-left: 5px;
    margin-top: 5px;
}
</style>