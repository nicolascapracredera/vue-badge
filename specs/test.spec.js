import CharacterDetail from '../src/components/CharacterDetail.vue';
import CharacterListing from '../src/components/CharacterListing.vue';
import Search from '../src/pages/Search.vue';
import Home from '../src/pages/Home.vue';
import { mount } from '@vue/test-utils';
import 'regenerator-runtime/runtime';

test('character detail zeros', () => {
    const wrapper = mount(CharacterDetail, {
        propsData: {
            character: {
                comics: {
                    available: 0
                },
                name: "Test Name",
                series: {
                    available: 0
                },
                stories: {
                    available: 0
                },
                thumbnail: {
                    path: "http://test-path-for-image.com",
                    extension: "jpg"
                }
            }
        }
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.text()).toContain("Appears in 0 comics, 0 series, and 0 stories.");
});

test('character detail numbers', () => {
    const wrapper = mount(CharacterDetail, {
        propsData: {
            character: {
                comics: {
                    available: 7
                },
                name: "Test Name",
                series: {
                    available: 234
                },
                stories: {
                    available: 543
                },
                thumbnail: {
                    path: "http://test-path-for-image.com",
                    extension: "jpg"
                }
            }
        }
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.text()).toContain("Appears in 7 comics, 234 series, and 543 stories.");
});

test('search wrapper check', () => {
    const searchWrapper = mount(Search);

    expect(searchWrapper.html()).toMatchSnapshot();
});

test('character listing button check', async () => {
    const wrapper = mount(CharacterListing);
    await wrapper.setData({
        characters: [
            {
                comics: {
                    available: 0
                },
                name: "Test Name 1",
                series: {
                    available: 0
                },
                stories: {
                    available: 0
                },
                thumbnail: {
                    path: "http://test-path-for-image.com",
                    extension: "jpg"
                }
            },
            {
                comics: {
                    available: 7
                },
                name: "Test Name 2",
                series: {
                    available: 234
                },
                stories: {
                    available: 543
                },
                thumbnail: {
                    path: "http://test-path-for-image.com",
                    extension: "jpg"
                }
            }
        ]
    });

    const firstCharacterButton = wrapper.find('button');

    expect(wrapper.text()).not.toContain("Appears in 0 comics, 0 series, and 0 stories.");

    await firstCharacterButton.trigger('click');

    expect(wrapper.text()).toContain("Appears in 0 comics, 0 series, and 0 stories.");

    expect(wrapper.html()).toMatchSnapshot();
    wrapper.destroy();
});