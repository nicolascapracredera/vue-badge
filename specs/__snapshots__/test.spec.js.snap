// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`character detail numbers 1`] = `
"<div>
  <h3>Detailed Character Info</h3>
  <div class=\\"card\\" style=\\"width: 18rem;\\"><img alt=\\"Character Photo\\" src=\\"http://test-path-for-image.com/portrait_incredible.jpg\\" class=\\"card-img-top\\">
    <div class=\\"card-body\\">
      <p>Name: Test Name</p>
      <p>Appears in 7 comics, 234 series, and 543 stories.</p>
      <p>Test Name was last modified on Invalid Date</p>
    </div>
  </div>
</div>"
`;

exports[`character detail zeros 1`] = `
"<div>
  <h3>Detailed Character Info</h3>
  <div class=\\"card\\" style=\\"width: 18rem;\\"><img alt=\\"Character Photo\\" src=\\"http://test-path-for-image.com/portrait_incredible.jpg\\" class=\\"card-img-top\\">
    <div class=\\"card-body\\">
      <p>Name: Test Name</p>
      <p>Appears in 0 comics, 0 series, and 0 stories.</p>
      <p>Test Name was last modified on Invalid Date</p>
    </div>
  </div>
</div>"
`;

exports[`character listing button check 1`] = `
"<div>
  <div id=\\"listing-container\\">
    <p>Total Results: 0</p>
    <div id=\\"character-listing\\">
      <div class=\\"btn-col\\"><button class=\\"btn btn-outline-primary character-name\\">
          Test Name 1
        </button><button class=\\"btn btn-outline-primary character-name\\">
          Test Name 2
        </button></div>
      <div class=\\"btn-col\\"></div>
      <!---->
    </div>
  </div>
  <div>
    <h3>Detailed Character Info</h3>
    <div class=\\"card\\" style=\\"width: 18rem;\\"><img alt=\\"Character Photo\\" src=\\"http://test-path-for-image.com/portrait_incredible.jpg\\" class=\\"card-img-top\\">
      <div class=\\"card-body\\">
        <p>Name: Test Name 1</p>
        <p>Appears in 0 comics, 0 series, and 0 stories.</p>
        <p>Test Name 1 was last modified on Invalid Date</p>
      </div>
    </div>
  </div>
</div>"
`;

exports[`search wrapper check 1`] = `
"<div id=\\"container\\">
  <p>Start typing in the box below to search for a Marvel character. Click on a character's name to view details about them.</p> <input type=\\"text\\" placeholder=\\"Type Character Name\\" class=\\"form-control col-4\\">
  <div>
    <div id=\\"listing-container\\">
      <p>Total Results: 0</p>
      <div id=\\"character-listing\\">
        <div class=\\"btn-col\\"></div>
        <div class=\\"btn-col\\"></div>
        <!---->
      </div>
    </div>
    <!---->
  </div>
</div>"
`;
